uarray/_uarray_dispatch.cxx

```diff
@@ -822,52 +829,133 @@ int Function_traverse(Function * self, visitproc visit, void * arg)
 }
 
 
+/** Break reference cycles when being GCed */
+int Function::clear(Function * self)
+{
+  self->extractor_.reset();
+  self->replacer_.reset();
+  self->def_args_.reset();
+  self->def_kwargs_.reset();
+  self->def_impl_.reset();
+  self->dict_.reset();
+  return 0;
+}
+
+
+/** Support for pickle.dump */
+PyObject * Function::getstate__(Function * self, PyObject * /*args*/)
+{
+  auto domain = py_ref::steal(
+    PyUnicode_FromStringAndSize(self->domain_key_.data(),
+                                self->domain_key_.size()));
+  if (!domain)
+    return nullptr;
+
+  auto res = py_make_tuple(self->extractor_,
+                           self->replacer_,
+                           domain,
+                           self->def_args_,
+                           self->def_kwargs_,
+                           self->def_impl_,
+                           self->dict_ ? self->dict_ : Py_None);
+
+  return res.release();
+}
+
+
+/** Support for pickle.load */
+PyObject * Function::setstate__(Function * self, PyObject * args)
+{
+  PyObject * state_tuple = nullptr;
+  if (!PyArg_ParseTuple(args, "O!", &PyTuple_Type, &state_tuple))
+    return nullptr;
+
+  if (PyTuple_GET_SIZE(state_tuple) != 7)
+  {
+    PyErr_SetString(PyExc_ValueError, "State tuple must have 7 entries");
+    return nullptr;
+  }
+
+  auto init_args = py_ref::steal(PyTuple_GetSlice(state_tuple, 0, 6));
+  if (!init_args)
+    return nullptr;
+
+  if (Function::init(self, init_args, nullptr) < 0)
+    return nullptr;
+
+  auto dict = py_ref::ref(PyTuple_GET_ITEM(state_tuple, 6));
+  if (dict == Py_None)
```

Comment:
Does the test hit both branches here?

Reply:
I don't think this can actually happen in normal usage because `generate_multimethod` calls `functools.update_wrapper` which adds attributes.

Reply:
I could add a test case that creates the `_Function` object directly though.

Reply:
Fair enough... Would it be worth it to add a test accessing the private class or removing the branch entirely?

Reply:
> I could add a test case that creates the _Function object directly though.

That’d be nice.

Reply:
Done